modules/local/cluster_file/main.nf

```diff
@@ -0,0 +1,44 @@
+process CLUSTER_FILE {
+    tag "Create cluster file for GAS call"
+    label 'process_single'
+
+    input:
+    val meta
+
+    output:
+    path("expected_clusters.txt"), emit: text
+
+    exec:
+    def outputLines = []
+
+    // Determine the maximum number of levels to set the header requirements for each pipeline run
+    int maxLevels = meta.collect { sample -> sample.address.split("\\$params.gm_delimiter").size() }.max() ?: 0
+
+    // Verify each sample is consistent with $maxLevels
+    meta.each { sample ->
+        int level = sample.address.split("\\$params.gm_delimiter").size()
+        if (level != maxLevels) {
+            error ("Inconsistent levels found: expected $maxLevels levels but found $level levels in ${sample.id}")
+        }
+    }
+
+    // Generate the header for the expected_clusters.txt file
+    def header = ["id", "address"] + (1..maxLevels).collect { "level_$it" }
+    outputLines << header.join("\t")
+
+    // Iterate over each sample in the meta list and pull the relevant information for the text file
+    meta.each { sample ->
+        def id = sample.id
+        def address = sample.address
+        def levels = address.split("\\$params.gm_delimiter")
+        def line = [id, address] + levels.collect { it.toString() } + (levels.size()..<maxLevels).collect { "" }
```

Comment:
Is the `(levels.size()..<maxLevels).collect { "" }` part padding out the line when levels are missing?

Reply:
Correct: `(levels.size()..<maxLevels).collect { "" }` pads out the line by adding empty strings for missing levels in the address.
**However**, since the script validates the consistency of levels across all samples and expects values in each level, this step is redundant and also potentially problematic...

So ... it has been removed here: [20f5522](https://github.com/phac-nml/gasnomenclature/pull/11/commits/20f55224e9f6d7af2eb6913288159ef981a2b263)

Although, now I am questioning whether gas-call can allow missing levels, so long as the maximum level is the same. If that is the case, the `nextflow_schema.json` will have to be updated to allow `meta.address` to have hashes...